pss/trojan_chunk.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pss
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"encoding/json"
+	"math/big"
+	"reflect"
+
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/storage"
+)
+
+// TODO: can we re-use some existing types here?
+type trojanHeaders struct {
+	span  []byte
+	nonce []byte
+}
+
+// TODO: can we re-use some existing types here?
+type trojanMessage struct {
+	length  []byte
+	topic   []byte
+	payload []byte
+	padding []byte
+}
+
+type trojanData struct {
+	trojanHeaders
+	trojanMessage // TODO: this should be encrypted
+}
+
+type trojanChunk struct {
+	address chunk.Address
+	trojanData
+}
+
+// newTrojanChunk creates a new trojan chunk structure for the given address and message
+func newTrojanChunk(address chunk.Address, message trojanMessage) (*trojanChunk, error) {
+	chunk := &trojanChunk{
+		address: address,
+		trojanData: trojanData{
+			trojanHeaders: newTrojanHeaders(),
+			trojanMessage: message,
+		},
+	}
+	// find nonce for chunk
+	if err := chunk.setNonce(); err != nil {
+		return nil, err
+	}
+	return chunk, nil
+}
+
+// newTrojanHeaders creates an empty trojan headers struct
+func newTrojanHeaders() trojanHeaders {
+	// TODO: what should be the value of this?
+	span := make([]byte, 8)
+	// create initial nonce
```

Comment:
why? and why do we want the nonce as a field at all?

Reply:
is the nonce not part of the trojan message?

Reply:
i've removed the nonce as part of the message. 

it is now a field used to put together a trojan chunk (which is of the `chunk.Chunk` type)